src/sidebar/store/use-store.js

```diff
@@ -6,6 +6,16 @@ import shallowEqual from 'shallowequal';
 import warnOnce from '../../shared/warn-once';
 import { useService } from '../util/service-context';
 
+/**
+ * @typedef {import("redux").Store} Store
+ */
+
+/**
+ * @callback StoreCallback
+ * @param {Store} store
+ * @return {any}
+ */
```

Comment:
I have a couple of recommendations here:

1. Rather than use `any` as the return type of `StoreCallback` you can make this function generic with `@template` and TS will infer it automatically. This will mean later that whenever we have `const aVar = useStore(store => store.something())` then TS will infer that `aVar` has whatever type `store.something()` returns.
2. I found a better pattern for dealing with refs that start off as `null` and then get set to a value later. If you do. `const aVar = useRef(/** @type {<the type>|null} */ (null))` then TS can infer everything else.

Here is the diff of that file with those two changes applied:

```diff
diff --git a/src/sidebar/store/use-store.js b/src/sidebar/store/use-store.js
index 475de6d2..0ca3e321 100644
--- a/src/sidebar/store/use-store.js
+++ b/src/sidebar/store/use-store.js
@@ -11,9 +11,10 @@ import { useService } from '../util/service-context';
  */
 
 /**
+ * @template T
  * @callback StoreCallback
  * @param {Store} store
- * @return {any}
+ * @return {T}
  */
 
 /**
@@ -40,7 +41,7 @@ import { useService } from '../util/service-context';
  *   }
  *
  * @template T
- * @param {StoreCallback} callback -
+ * @param {StoreCallback<T>} callback -
  *   Callback that receives the store as an argument and returns some state
  *   and/or actions extracted from the store.
  * @return {T} - The result of `callback(store)`
@@ -50,10 +51,10 @@ export default function useStore(callback) {
 
   // Store the last-used callback in a ref so we can access it in the effect
   // below without having to re-subscribe to the store when it changes.
-  const lastCallback = useRef();
+  const lastCallback = useRef(/** @type {typeof callback|null} */ (null));
   lastCallback.current = callback;
 
-  const lastResult = useRef();
+  const lastResult = useRef(/** @type {T|undefined} */(undefined));
   lastResult.current = callback(store);
 
   // Check for a performance issue caused by `callback` returning a different
@@ -75,13 +76,13 @@ export default function useStore(callback) {
   // and re-render the component if the result changed.
   useEffect(() => {
     function checkForUpdate() {
-      const result = /** @type {StoreCallback} */ (lastCallback.current)(store);
+      const result = lastCallback.current(store);
       if (shallowEqual(result, lastResult.current)) {
         return;
       }
       lastResult.current = result;
       // Force this function to ignore parameters and just force a store update.
-      /** @type {()=>any} */ (forceUpdate)();
+      forceUpdate(undefined);
     }
 
     // Check for any changes since the component was rendered.
@@ -94,5 +95,5 @@ export default function useStore(callback) {
     return unsubscribe;
   }, [forceUpdate, store]);
 
-  return /** @type {T} */ (lastResult.current);
+  return lastResult.current;
 }
```

Reply:
Do you any preference between these two implementations?

`const lastCallback = useRef(/** @type {StoreCallback<T>|null} */ (null));`
`const lastCallback = useRef(/** @type {typeof callback|null} */ (null));`

For me, I would probably prefer asserting the type directly rather than using the via a local variable. But perhaps there is a distinction here I'm missing?

Reply:
I would guess the first one might generate slightly easier to read error messages and also save a human reader the need to go and look at what `callback` is. I think I'd agree to go with that unless the second version was significantly shorter in a certain context.